### PR TITLE
[DUOS-1972][risk=medium] V3 dataset registration endpoint stub

### PR DESCRIPTION
**Pull request by rushtong**
## Addresses
Partially addresses https://broadworkbench.atlassian.net/browse/DUOS-1972

Marked as **Medium** risk due to the optional file upload.

New endpoint for registering a dataset submission. API takes a specific schema and generates N (minimum 1) datasets from the provided data. 

### This PR covers:
* API Endpoint
* API Documentation
  * Includes some minor category renaming and v2 deprecation
* Tests
* Schema validation

Future PRs will cover the creation of files and datasets from the provided schema.

All classes in `org.broadinstitute.consent.http.models.dataset_registration_v1` are generated via [jsonschema2pojo](https://github.com/joelittlejohn/jsonschema2pojo). These required some manual tweaks to work correctly with our classes, in particular `Double`s need to be `Integer`s and enums had case sensitivity issues.

### Example swagger

Screen Shot 2022-09-29 at 4 00 51 PM

### Testing

- Spin up the branch
- Construct sample registration
- Authenticate as admin or data submitter
- Post with and without file contents
- Putting the debugger on the endpoint should show that all content is being parsed and validated

----
Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

- Label PR with a Jira ticket number and include a link to the ticket
- Label PR with a security risk modifier [no, low, medium, high]
- PR describes scope of changes
- Get a minimum of one thumbs worth of review, preferably two if enough team members are available
- Get PO sign-off for all non-trivial UI or workflow changes
- Verify all tests go green
- Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/resources/dataset-registration-schema_v1.json [hunk 1]
```diff
@@ -355,6 +356,12 @@
           "label": "Publication Moratorium (MOR)",
           "description": "Publication Moratorium (MOR)"
         },
+        "morDate": {
+          "type": "string",
+          "format": "date",
+          "label": "Publication Moratorium Date (MOR)",
+          "description": "Publication Moratorium Date (MOR)"
+        },
```
Comment (connorlbark) [inline src/main/resources/dataset-registration-schema_v1.json:365]: right now in the code merged into the UI, the date is saved in the `mor` field, and is undefined if the user does not check it. easy fix, but would need to be changed

Comment (rushtong) [reply src/main/resources/dataset-registration-schema_v1.json:365]: Yes, I think that was a sort of workaround. When I reviewed the mocks, we needed two fields, a boolean for the field and then a date value for an actual date. If the user unselected/toggled the boolean option between saves, we could remember the original date selection.

Screen Shot 2022-09-29 at 4 05 24 PM

Comment (rushtong) [reply src/main/resources/dataset-registration-schema_v1.json:365]: Also, there was a slack conversation that clarified this to some degree.

Comment (connorlbark) [reply src/main/resources/dataset-registration-schema_v1.json:365]: ahhh yeah, I think I achieved remembering it by keeping state outside of the formData, in a react state variable. I could see arguments for either, though

File: src/main/java/org/broadinstitute/consent/http/models/dataset_registration_v1/ConsentGroup.java [hunk 2]
```diff
@@ -0,0 +1,606 @@
...
+    @JsonProperty("mor")
+    @JsonPropertyDescription("Publication Moratorium (MOR)")
+    private Boolean mor;
+    /**
+     * Publication Moratorium Date (MOR)
+     *
+     */
+    @JsonProperty("morDate")
+    @JsonPropertyDescription("Publication Moratorium Date (MOR)")
+    private String morDate;
```
Comment (connorlbark) [inline src/main/java/org/broadinstitute/consent/http/models/dataset_registration_v1/ConsentGroup.java:135]: should this (and other dates) be parsed to date / validated?

File: src/main/java/org/broadinstitute/consent/http/models/dataset_registration_v1/DatasetRegistrationSchemaV1.java [hunk 3]
```diff
@@ -0,0 +1,1582 @@
...
+    /**
+     * Public Visibility of this study
+     * (Required)
+     *
+     */
+    @JsonProperty("publicVisibility")
+    @JsonPropertyDescription("Public Visibility of this study")
+    private Boolean publicVisibility;
+    @JsonProperty("nihAnvilUse")
+    private List<NihAnvilUse> nihAnvilUse = new ArrayList<NihAnvilUse>();
```
Comment (connorlbark) [inline src/main/java/org/broadinstitute/consent/http/models/dataset_registration_v1/DatasetRegistrationSchemaV1.java:165]: needs comment ?

File: src/main/java/org/broadinstitute/consent/http/util/JsonSchemaUtil.java [hunk 4]
```diff
@@ -45,4 +47,19 @@ public boolean isValidSchema_v1(String datasetRegistrationInstance) {
       return false;
     }
   }
+
+  public DatasetRegistrationSchemaV1 deserializeDatasetRegistration(String datasetRegistrationInstance) {
+    try {
+      String schemaString = IOUtils.resourceToString("/dataset-registration-schema_v1.json", Charset.defaultCharset());
```
Comment (connorlbark) [inline src/main/java/org/broadinstitute/consent/http/util/JsonSchemaUtil.java]: would it make sense / be possible to store this in memory so it doesn't have to reload every time someone makes a dataset?

File: src/main/java/org/broadinstitute/consent/http/service/DatasetService.java [hunk 5]
```diff
@@ -51,15 +55,17 @@ public class DatasetService {
...
     private final UseRestrictionConverter converter;
 
     @Inject
     public DatasetService(ConsentDAO consentDAO, DataAccessRequestDAO dataAccessRequestDAO, DatasetDAO dataSetDAO,
-                          UserRoleDAO userRoleDAO, UseRestrictionConverter converter) {
+                          GCSService gcsService, UserRoleDAO userRoleDAO, UseRestrictionConverter converter) {
         this.consentDAO = consentDAO;
         this.dataAccessRequestDAO = dataAccessRequestDAO;
         this.datasetDAO = dataSetDAO;
+        this.gcsService = gcsService;
```
Comment (connorlbark) [inline src/main/java/org/broadinstitute/consent/http/service/DatasetService.java:68]: is gcsService needed? will it be needed in the follow up PR?

File: src/test/java/org/broadinstitute/consent/http/util/JsonSchemaUtilTest.java [hunk 6]
```diff
@@ -1,46 +1,71 @@
...
+    assertFalse(instance.getDataTypes().isEmpty());
+    assertFalse(instance.getFileTypes().isEmpty());
+    assertNotNull(instance.getPhenotypeIndication());
+    assertNotNull(instance.getSpecies());
+    assertNotNull(instance.getPiName());
+    assertNotNull(instance.getDataSubmitterUserId());
+    assertFalse(instance.getDataCustodianEmail().isEmpty());
+    assertNotNull(instance.getPublicVisibility());
+    assertNotNull(instance.getDataAccessCommitteeId());
+    assertFalse(instance.getConsentGroups().isEmpty());
```
Comment (connorlbark) [inline src/test/java/org/broadinstitute/consent/http/util/JsonSchemaUtilTest.java:75]: should we have more comprehensive testing of e.g., enum parsing / other custom types?

**Review by connorlbark — COMMENTED** [review]
looking very cool!! exciting to see how this tighten up requirements & make registration more robust. just a couple comments!

File: src/main/java/org/broadinstitute/consent/http/models/dataset_registration_v1/ConsentGroup.java (hunk 2 above)
Comment (rushtong) [reply src/main/java/org/broadinstitute/consent/http/models/dataset_registration_v1/ConsentGroup.java:135]: These classes are auto-generated. If we re-generate them, we need to repeat any changes we made. I had to do some already, and want to minimize that as much as possible. Additionally, the schema validation is responsible for parsing/validating the date format. For those reasons, I don't want to add more to these classes than is necessary.

File: src/main/java/org/broadinstitute/consent/http/models/dataset_registration_v1/DatasetRegistrationSchemaV1.java (hunk 3 above)
Comment (rushtong) [reply src/main/java/org/broadinstitute/consent/http/models/dataset_registration_v1/DatasetRegistrationSchemaV1.java:165]: 👍🏽 This is because I forgot to add a description in the schema for this field.

File: src/main/java/org/broadinstitute/consent/http/service/DatasetService.java (hunk 5 above)
Comment (rushtong) [reply src/main/java/org/broadinstitute/consent/http/service/DatasetService.java:68]: It will be needed to store the upload file and yes, will be used in the next PR.

File: src/test/java/org/broadinstitute/consent/http/util/JsonSchemaUtilTest.java (hunk 6 above)
Comment (rushtong) [reply src/test/java/org/broadinstitute/consent/http/util/JsonSchemaUtilTest.java:75]: We do need more testing - am hoping to do more in a future PR. I anticipate some changes as we integrate the UI and the API, so am holding off on that for now.

File: src/main/java/org/broadinstitute/consent/http/util/JsonSchemaUtil.java (hunk 4 above)
Comment (rushtong) [reply src/main/java/org/broadinstitute/consent/http/util/JsonSchemaUtil.java]: Excellent idea ... will do 👍🏽

File: src/main/java/org/broadinstitute/consent/http/models/dataset_registration_v1/ConsentGroup.java (hunk 2 above)
Comment (connorlbark) [reply src/main/java/org/broadinstitute/consent/http/models/dataset_registration_v1/ConsentGroup.java:135]: got it, makes sense. As long as it is validating the format, this is good with me.

File: src/main/resources/assets/paths/datasetV3.yaml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+          type: object
+          properties:
+            dataset:
+              description: Required dataset registration json schema instance
+              type: object
+              $ref: '../schemas/DatasetRegistrationSchemaV1.yaml'
+            file:
+              description: Optional alternative data sharing plan
+              type: string
+              format: binary
```
Comment (connorlbark) [inline src/main/resources/assets/paths/datasetV3.yaml:22]: trying to test out this endpiont, in swagger, this seems to show the dataset as the request body for some reason:
image
It seems to be totally wrong?? I can't figure out why.

Comment (connorlbark) [reply src/main/resources/assets/paths/datasetV3.yaml:22]: There's not multipart/form-data option in the dropdown.

Comment (rushtong) [reply src/main/resources/assets/paths/datasetV3.yaml:22]: I'm not sure you're looking at the latest version. I see something pretty different. I do see some browser caching issues, but the endpoint looks correct to me:

Screen Shot 2022-10-03 at 12 28 56 PM

Comment (connorlbark) [reply src/main/resources/assets/paths/datasetV3.yaml:22]: 🤔 curious, I tried emptying my cache (I have had that issue before, but I thought I had fixed it when I emptied the cache). Let me try again..

Comment (connorlbark) [reply src/main/resources/assets/paths/datasetV3.yaml:22]: hmmmmm so I cleared the website's cache and the entire browser's cache, and it still has the old information up, which makes no sense because the information here does not relfect what is in swagger. 🤔  very odd

Comment (connorlbark) [reply src/main/resources/assets/paths/datasetV3.yaml:22]: OHHH I WAS ON THE WRONG BRANCH, DUOS-1972 instead of DUOS-1972-resource-stub.

**Comment by connorlbark** [comment]
validation functionally works very well!!! looking very cool.
- enums all validated properly
- types validated properly
- emails validated properly
- dates did NOT validate properly - `asdf-10-20` was a valid date.

in addition - would a more specific schema error message be worth investigating?

**Comment by rushtong** [comment]
> * dates did NOT validate properly - `asdf-10-20` was a valid date.
> 
> in addition - would a more specific schema error message be worth investigating?

@connorlbark 
I added custom date formatting since the validation library doesn't seem to support it as documented. Will look into more error reporting as well.

**Review by connorlbark — APPROVED** [review]
Date validation now works, file upload works, everything seems ok !

**Review by quazi-h — APPROVED** [review]
The new endpoint appears to be functioning correctly. I was able to use a remote debugger on the new DatasetResource endpoint and observed the elements from the registration schema provided as input being parsed correctly. Also confirmed that validation is working and received 400 "Invalid Schema" errors when providing invalid inputs.
